test/FunctionalTests/Microsoft.OData.Core.Tests/ODataWriterValidationTests.cs

```diff
@@ -0,0 +1,3356 @@
+            {
+                var exception = await Assert.ThrowsAsync<ODataException>(async () =>
+                {
+                    var tempCustomerResource = new ODataResource
+                    {
+                        TypeName = "NS.Customer",
+                        Properties = new List<ODataProperty>
+                        {
+                            new ODataProperty { Name = "Id", Value = 1 },
+                            new ODataProperty { Name = "Emails", Value = "sue@temp.com" } // Emails is an collection property
+                        }
+                    };
+
+                    var resourceWriter = await messageWriter.CreateODataResourceWriterAsync(this.customersEntitySet, this.customerEntityType);
+
+                    await resourceWriter.WriteStartAsync(tempCustomerResource);
+                    await resourceWriter.WriteEndAsync();
+                });
+
+                Assert.Equal(
+                    Strings.ValidationUtils_NonPrimitiveTypeForPrimitiveValue("Collection(Edm.String)"),
+                    exception.Message);
+            }
+        }
+
+        [Fact]
+        public async Task ValidationUtils_PropertiesMustNotContainReservedCharsAsync()
+        {
+            IODataRequestMessage requestMessage = new InMemoryMessage { Stream = this.stream };
+
+            await using (var messageWriter = new ODataMessageWriter(requestMessage, this.messageWriterSettings, this.model))
+            {
+                var exception = await Assert.ThrowsAsync<ODataException>(async () =>
+                {
+                    var tempCustomerResource = new ODataResource
+                    {
+                        TypeName = "NS.Customer",
+                        Id = new Uri("http://tempuri.org/Customers(1)"),
+                        Properties = new List<ODataProperty>
+                        {
+                            new ODataProperty { Name = "Id", Value = 1 },
+                            new ODataProperty { Name = "Customer.Name", Value = "Sue" },
+                        }
+                    };
+
+                    var resourceWriter = await messageWriter.CreateODataResourceWriterAsync(this.customersEntitySet, this.customerEntityType);
+
+                    await resourceWriter.WriteStartAsync(tempCustomerResource);
+                    await resourceWriter.WriteEndAsync();
+                });
+
+                Assert.Equal(
+                    Strings.ValidationUtils_PropertiesMustNotContainReservedChars("Customer.Name", "':', '.', '@'"),
+                    exception.Message);
+            }
+        }
+
+        [Theory]
+        [InlineData("")]
+        [InlineData(null)]
+        public async Task WriterValidationUtils_PropertiesMustHaveNonEmptyNameAsync(string propertyName)
+        {
+            IODataRequestMessage requestMessage = new InMemoryMessage { Stream = this.stream };

```

Comment:
It seems this line is for all test case, maybe it's better to make it as instance field and You don't need to repeat it in every unit test.

Reply:
@xuzhg Refactored such that all tests use helper methods that contain the shared logic